hibernate-core/src/test/java/org/hibernate/test/sql/hand/query/NativeSQLQueriesTest.java

```diff
@@ -0,0 +1,883 @@
+					session.persist( jboss );
+
+					// now query on Employment, this should not cause an auto-flush
+					session.createNativeQuery( getEmploymentSQL() ).addSynchronizedQuerySpace( "ABC" ).list();
+					assertEquals( 0, session.getSessionFactory().getStatistics().getEntityInsertCount() );
+
+					// now try to query on Employment but this time add Organization as a synchronized query space...
+					session.createNativeQuery( getEmploymentSQL() ).addSynchronizedEntityClass( Organization.class ).list();
+					assertEquals( 1, session.getSessionFactory().getStatistics().getEntityInsertCount() );
+
+					// clean up
+					session.delete( jboss );
+				}
+		);
+	}
+
+	@Test
+	public void testSQLQueryInterface(SessionFactoryScope scope) {
+		Organization ifa = new Organization("IFA");
+		Organization jboss = new Organization("JBoss");
+		Person gavin = new Person("Gavin");
+		Employment emp = new Employment(gavin, jboss, "AU");
+
+		scope.inTransaction(
+				session -> {
+
+					session.persist(ifa);
+					session.persist(jboss);
+					session.persist(gavin);
+					session.persist(emp);
+
+					List l = session.createNativeQuery( getOrgEmpRegionSQL() )
+							.addEntity("org", Organization.class)
+							.addJoin("emp", "org.employments")
+							.addScalar("regionCode", StandardBasicTypes.STRING )
+							.list();
+					assertEquals( 2, l.size() );
+
+					l = session.createNativeQuery( getOrgEmpPersonSQL() )
+							.addEntity("org", Organization.class)
+							.addJoin("emp", "org.employments")
+							.addJoin("pers", "emp.employee")
+							.list();
+					assertEquals( l.size(), 1 );
+				}
+		);
+
+		scope.inTransaction(
+				session -> {
+					List l = session.createNativeQuery( "select {org.*}, {emp.*} " +
+																"from ORGANIZATION org " +
+																"     left outer join EMPLOYMENT emp on org.ORGID = emp.EMPLOYER, ORGANIZATION org2" )
+							.addEntity("org", Organization.class)
+							.addJoin("emp", "org.employments")
+							.setResultTransformer( DistinctRootEntityResultTransformer.INSTANCE )
+							.list();
+					assertEquals( l.size(), 2 );
+				}
+		);
+
+		scope.inTransaction(
+				session -> {
+					session.delete(emp);
+					session.delete(gavin);
+					session.delete(ifa);
+					session.delete(jboss);
+				}
+		);
+	}
+
+	@Test
+	public void testResultSetMappingDefinition(SessionFactoryScope scope) {
+		Organization ifa = new Organization("IFA");
+		Organization jboss = new Organization("JBoss");
+		Person gavin = new Person("Gavin");
+		Employment emp = new Employment(gavin, jboss, "AU");
+
+		scope.inTransaction(
+				session -> {
+					session.persist(ifa);
+					session.persist(jboss);
+					session.persist(gavin);
+					session.persist(emp);
+
+					List l = session.createNativeQuery( getOrgEmpRegionSQL(), "org-emp-regionCode" ).list();
+					assertEquals( l.size(), 2 );
+
+					l = session.createNativeQuery( getOrgEmpPersonSQL(), "org-emp-person" ).list();
+					assertEquals( l.size(), 1 );
+
```

Comment:
It is common defect to mess up the parameter order of assertEquals(). It is not a big deal when it succeeds, but would generate confusing message when it fails.